### PR TITLE
Action for running tests in parallel

### PR DESCRIPTION
**Pull request by petrovic-d**
New action for running tests in parallel. It can be invoked via command: `nbls.run.test.parallel`
Command support specifying modules for which tests will be ran in parallel. It's up to client to pass the module list, so by default it will run without passed modules. 
To test it with module list, it needs to be modified like this:
in `java/java.lsp.server/vscode/src/extension.ts` file
modify line `835` to:
`testAdapter?.run(new vscode.TestRunRequest(), new vscode.CancellationTokenSource().token, true, ["module1", "module2"]);`

### CODE REVIEW

File: java/maven/src/org/netbeans/modules/maven/execute/MavenCommandLineExecutor.java [hunk 1]
```diff
@@ -445,6 +445,18 @@ private static List<String> createMavenExecutionCommand(RunConfig config, Constr
                 LOGGER.log(Level.FINE, "Could not canonicalize " + basedir, x);
             }
         }
+    
+        for (Map.Entry<? extends String, ? extends String> entry : config.getOptions().entrySet()) {
+            String key = entry.getKey();
+            String value = quote2apos(entry.getValue());
+            if (MavenCommandLineOptions.optionRequiresValue(key) && (value.isBlank() || value.equals("${" + key + "}"))) { //NOI18N
```
Comment (sdedic) [inline java/maven/src/org/netbeans/modules/maven/execute/MavenCommandLineExecutor.java:452]: why the check for `${optionName}` ?
Consider to print a warning visibly to the user: if trying to pass no value for an option that is known to have one, it's a bug somewhere in the code.

Maybe I am picky, but `isBlank()` is true also for "  " (two spaces); I wouldn't equal that to a no-value. Consider value quoting (see proprerties below).

File: java/maven/src/org/netbeans/modules/maven/execute/MavenCommandLineOptions.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+import java.util.List;
+
+/**
+ *
+ * @author Dusan Petrovic
+ */
+public class MavenCommandLineOptions {
+    
+    private static final List<String> OPTIONS_WITH_VALUES = List.of(
+            "file",
```
Comment (sdedic) [inline java/maven/src/org/netbeans/modules/maven/execute/MavenCommandLineOptions.java:30]: Also
- builder
-  color
- define
- global-toolchains

Pls.order alphabetically in source, it is easier to find an option when reading :) I'd use `Set.of`, faster `contains` lookups.

File: java/maven/src/org/netbeans/modules/maven/execute/cmd/ExecMojo.java [hunk 3]
```diff
@@ -155,4 +169,24 @@ public String getImplementationClass() {
...
     }
+
+    public @CheckForNull Project findProject() {
+        if (currentProjectLocation != null) {
+            FileObject fo = FileUtil.toFileObject(currentProjectLocation);
+            if (fo != null) {
+                try {
+                    return ProjectManager.getDefault().findProject(fo);
+                } catch (IOException ex) {
+                    //Exceptions.printStackTrace(ex);
```
Comment (sdedic) [inline java/maven/src/org/netbeans/modules/maven/execute/cmd/ExecMojo.java:180]: Why this should not be logged ?

File: java/maven/src/org/netbeans/modules/maven/execute/model/NetbeansActionMapping.java [hunk 4]
```diff
@@ -264,6 +290,16 @@ public void removePackaging(String string)
         getPackagings().remove( string );
     } //-- void removePackaging(String) 
 
+    /**
+     * Method removeOptions.
+     * 
+     * @param string
+     */
+    public void removeOptions(String string)
+    {
```
Comment (sdedic) [inline java/maven/src/org/netbeans/modules/maven/execute/model/NetbeansActionMapping.java:299]: singular `removeOption` ?

**Review by sdedic — CHANGES_REQUESTED** [review]

File: java/java.lsp.server/vscode/src/testAdapter.ts [hunk 5]
```diff
@@ -198,19 +213,41 @@ export class NbTestAdapter {
...
     }
 
+    calculateStateFor(testItem: TestItem): 'passed' | 'failed' | 'skipped' | 'errored' {
+        let passed: number = 0;
+        testItem.children.forEach(item => {
+            const state = this.suiteStates.get(item);
+            if (state === 'enqueued' || state === 'failed') return state;
+            if (state === 'passed') passed++;
+        })
+        if (passed > 0) return 'passed';
```
[the remark this answers is not in the document]
Comment (petrovic-d) [reply java/java.lsp.server/vscode/src/testAdapter.ts:230]: This is used to calculate end state for the module node - https://github.com/apache/netbeans/blob/a3b539b8d2cd416f5ca54c800d6df2ca3bb590df/java/java.lsp.server/vscode/src/testAdapter.ts#L216, thats why there is no `started` state

File: java/gradle.test/src/org/netbeans/modules/gradle/test/GradleTestProgressListener.java [hunk 6]
```diff
@@ -241,11 +252,38 @@ private void caseFinish(TestFinishEvent evt, JvmTestOperationDescriptor op) {
...
+                }
+            }
+        }
+        return project;
+    }
+
+    private static String getSessionKey(OperationDescriptor op) {
+        String id = "";
+        for (OperationDescriptor descriptor = op; descriptor != null; descriptor = descriptor.getParent()) {
+            id = descriptor.getName();
```
[the remark this answers is not in the document]
Comment (dbalek) [reply java/gradle.test/src/org/netbeans/modules/gradle/test/GradleTestProgressListener.java:282]: IIRC- yes

File: java/gradle.test/src/org/netbeans/modules/gradle/test/GradleTestProgressListener.java [hunk 7]
```diff
@@ -241,11 +259,38 @@ private void caseFinish(TestFinishEvent evt, JvmTestOperationDescriptor op) {
                 }
 
             }
-            runningTests.remove(getTestOpKey(op));
         }
 
     }
 
+    private static String GRADLE_TEST_RUN = "Gradle Test Run :";
```
Comment (sdedic) [inline java/gradle.test/src/org/netbeans/modules/gradle/test/GradleTestProgressListener.java:266]: nitpick: make final :), `// NOI18N`

File: java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/debugging/launch/NbLaunchDelegate.java [hunk 8]
```diff
@@ -76,6 +76,7 @@
 import org.netbeans.modules.nativeimage.api.debug.StartDebugParameters;
 import org.netbeans.spi.project.ActionProgress;
 import org.netbeans.spi.project.ActionProvider;
+import org.netbeans.spi.project.ContainedProjectFilter;
```
Comment (sdedic) [inline java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/debugging/launch/NbLaunchDelegate.java:79]: bump dependency version for project.api

File: java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/progress/ModuleInfo.java [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.netbeans.modules.java.lsp.server.progress;
+
+/**
+ *
+ * @author dusanpetrovic
+ */
+public class ModuleInfo {
```
Comment (sdedic) [inline java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/progress/ModuleInfo.java:25]: final ?

File: java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/progress/TestProgressHandler.java [hunk 10]
```diff
@@ -144,4 +149,20 @@ private String statusToState(Status status) {
                 throw new IllegalStateException("Unexpected testsuite status: " + status);
         }
     }
+
+    private static Pattern PATTERN = Pattern.compile("Gradle Test Run :(.*):test");
```
Comment (sdedic) [inline java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/progress/TestProgressHandler.java:153]: make final, // NOI18N

File: java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/protocol/WorkspaceServiceImpl.java [hunk 11]
```diff
@@ -805,6 +808,13 @@ public CompletableFuture<Object> executeCommand(ExecuteCommandParams params) {
         throw new UnsupportedOperationException("Command not supported: " + params.getCommand());
     }
     
+    private String getModuleTestPath(Project project) {
+        if (project == null) {
+            return null;
+        }
+        return Paths.get(project.getProjectDirectory().getPath(), "src", "test", "java").toString();
```
Comment (sdedic) [inline java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/protocol/WorkspaceServiceImpl.java:815]: similar for the usage of Source API - whatever decision is taken.

File: java/maven/src/org/netbeans/modules/maven/ActionProviderImpl.java [hunk 12]
```diff
@@ -132,6 +132,7 @@ public class ActionProviderImpl implements ActionProvider {
         "javadoc", //NOI18N
         COMMAND_TEST,
         COMMAND_TEST_SINGLE,
+        COMMAND_TEST_PARALLEL,
```
Comment (sdedic) [inline java/maven/src/org/netbeans/modules/maven/ActionProviderImpl.java:135]: bump dependency version on project.api

File: java/maven/src/org/netbeans/modules/maven/api/execute/RunConfig.java [hunk 13]
```diff
@@ -79,6 +79,28 @@ public interface RunConfig {
 
     String getActionName();
     
+    /**
+     * Options/switches passed to maven.
+     * @return a read-only copy of the current maven options
+     * @since 2.166
+     */
+    @NonNull Map<? extends String,? extends String> getOptions();
```
Comment (sdedic) [inline java/maven/src/org/netbeans/modules/maven/api/execute/RunConfig.java:87]: Friend API change; probably add `apichanges.xml` entry for this + bump version of the module.

File: java/maven/src/org/netbeans/modules/maven/ActionProviderImpl.java (hunk 12 above)
Comment (neilcsmith-net) [reply java/maven/src/org/netbeans/modules/maven/ActionProviderImpl.java:135]: Just use current.  They're automatically bumped between each NetBeans release.

File: java/maven/src/org/netbeans/modules/maven/api/execute/RunConfig.java (hunk 13 above)
Comment (petrovic-d) [reply java/maven/src/org/netbeans/modules/maven/api/execute/RunConfig.java:87]: I already added entry for `options`

File: ide/projectapi/src/org/netbeans/spi/project/ContainedProjectFilter.java [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+ * under the License.
+ */
+package org.netbeans.spi.project;
+
+import java.util.Collections;
+import java.util.List;
+import org.netbeans.api.project.Project;
+
+/**
+ *
```
[the remark this answers is not in the document]
Comment (petrovic-d) [reply ide/projectapi/src/org/netbeans/spi/project/ContainedProjectFilter.java:26]: Well, we can introduce `filterMode`, which would be filled depending on the list size?

File: java/maven/nbproject/org-netbeans-modules-maven.sig
Comment (neilcsmith-net) [inline java/maven/nbproject/org-netbeans-modules-maven.sig:1]: Please remove the changes to this file.  Sigfiles are generated after each release.  Changes during delivery phase complicate that.

**Review by neilcsmith-net — CHANGES_REQUESTED** [review]

File: java/maven/src/org/netbeans/modules/maven/api/execute/RunConfig.java (hunk 13 above)
Comment (sdedic) [reply java/maven/src/org/netbeans/modules/maven/api/execute/RunConfig.java:87]: Sorry, I missed it in the diff.

**Comment by sdedic** [comment]
@neilcsmith-net please advise how to fix build failure https://github.com/apache/netbeans/actions/runs/12162766052/job/33936238338?pr=7979 related to signature check. There's a (formally) incompatible change on friend-only API.

File: java/maven.junit/src/org/netbeans/modules/maven/junit/JUnitOutputListenerProvider.java [hunk 15]
```diff
@@ -227,7 +218,7 @@ private boolean isSurefireRunningInParallel() {
     }
     
     private static final String SECONDS_REGEX = "s(?:ec(?:ond)?(?:s|\\(s\\))?)?"; //NOI18N
-    private static final String TESTSUITE_STATS_REGEX = "Tests run: +([0-9]+), +Failures: +([0-9]+), +Errors: +([0-9]+), +Skipped: +([0-9]+), +Time elapsed: +(.+)" + SECONDS_REGEX + " - in (.*)";
+    private static final String TESTSUITE_STATS_REGEX = "Tests run: +([0-9]+), +Failures: +([0-9]+), +Errors: +([0-9]+), +Skipped: +([0-9]+), +Time elapsed: +(.+)" + SECONDS_REGEX + " -+ in (.*)";
```
Comment (mbien) [inline java/maven.junit/src/org/netbeans/modules/maven/junit/JUnitOutputListenerProvider.java:221]: was the `+` intended?

**Review by mbien — COMMENTED** [review]
was this tested with mvnd? mvnd builds multithreaded by default, mvn builds singlethreaded by default (without extra flags).

File: java/maven.junit/src/org/netbeans/modules/maven/junit/JUnitOutputListenerProvider.java (hunk 15 above)
Comment (dbalek) [reply java/maven.junit/src/org/netbeans/modules/maven/junit/JUnitOutputListenerProvider.java:221]: Yes. `--` appear in the output.

**Comment by neilcsmith-net** [comment]
@sdedic sorry, I'd missed that the addition was incompatible.  Am not involved with releases at the moment, and only commented after a glance at this to try and save @ebarboni some headaches post NB24.

_If_ this has to go in as is, then either wait for the API sigs to merge in a week or so and rebase, manually patch the 3 required lines into the sig file (which should pass then?), or merge and be prepared to help Eric with the API signature integration post release.

OTOH, either the test is relevant or not?  Add the methods as default on the interface, as you did in the last change to that file to not require changes to pass tests?  Or lets just stop having sigtests for friend APIs as they're irrelevant - the build will pick up internal breakage, and we don't support external usage.

Our process for handling sig tests for releases could do with a review anyway.  I have some ideas how to improve that, but we have the one we have for now.

**Comment by petrovic-d** [comment]
@mbien I tried with `mvnd` and it ran all tests successfully with this action

File: java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/progress/TestProgressHandler.java [hunk 16]
```diff
@@ -154,7 +155,14 @@ private String statusToState(Status status) {
...
-        String modulePath = project != null ? project.getProjectDirectory().getPath() : null;
+        String modulePath = getModuleTestPath(project);
         return new ModuleInfo(moduleName, modulePath);
     }
+    
+    private static String getModuleTestPath(Project project) {
+        if (project == null) {
+            return null;
+        }
+        return Paths.get(project.getProjectDirectory().getPath(), "src", "test", "java").toString();
```
[the remark this answers is not in the document]
Comment (lahodaj) [reply java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/progress/TestProgressHandler.java:166]: I agree that hardcoding `src/test/java` is not ideal. I wonder how this is show is this visible in the UI? I tried to find out, but so far, I wasn't successful.

[the remark this answers is not in the document]
Comment (petrovic-d) [reply java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/progress/TestProgressHandler.java:166]: @lahodaj This is not visible in the UI. The purpose of it is to run tests for a single module (at least that's the idea). This way we know the path to the module tests on the client side. It's a bit coupled to this [PR](https://github.com/apache/netbeans/pull/8021) which fixes single test runs

**Comment by petrovic-d** [comment]
@ebarboni When can we expect sigs to be generated for the release, since this PR should (probably) introduce changes to .sig?

**Comment by neilcsmith-net** [comment]
@petrovic-d follow #8034

File: java/java.source.base/src/org/netbeans/api/java/source/SourceUtils.java [hunk 17]
```diff
@@ -1477,10 +1478,11 @@ public static void forceSource(CompilationController cc, FileObject file) {
      *
      * @param info the ClasspathInfo used to resolve
      * @param relativePath input source file path relative to the corresponding source root
+     * @param nestedClass nested class which name is searched
      * @return class name for the corresponding input source file
      * @since 2.73
      */
-    public static String classNameFor(ClasspathInfo info, String relativePath) {
+    public static String classNameFor(ClasspathInfo info, String relativePath, NestedClass nestedClass) {
```
Comment (sdedic) [inline java/java.source.base/src/org/netbeans/api/java/source/SourceUtils.java:1485]: Sorry I missed this one. Paperwork for this API addition is missing - `apichanges.xml` entry + update version for  dependency on `java.source.base' in maven + gradle modules which depend on this new API

File: java/maven/src/org/netbeans/modules/maven/api/execute/RunConfig.java [hunk 18]
```diff
@@ -79,6 +79,28 @@ public interface RunConfig {
 
     String getActionName();
     
+    /**
+     * Options/switches passed to maven.
+     * @return a read-only copy of the current maven options
+     * @since 2.166
```
Comment (sdedic) [inline java/maven/src/org/netbeans/modules/maven/api/execute/RunConfig.java:85]: The current master's version of java/maven is 2.166 as well, this should be increased, 2.166 is over 2 months old already.

File: ide/projectapi/src/org/netbeans/spi/project/SingleMethod.java [hunk 19]
```diff
@@ -53,6 +54,37 @@ public SingleMethod(FileObject file, String methodName) {
...
+     *
+     * @param file file to be kept in the object
+     * @param methodName name of a method inside the file     
+     * @param nestedClass nested class containing the method
+     * 
+     * @exception  java.lang.IllegalArgumentException
+     *             if the nested class name is {@code null}
+     * @since 1.99
+     */
+    public SingleMethod(FileObject file, String methodName, NestedClass nestedClass) {
```
Comment (sdedic) [inline ide/projectapi/src/org/netbeans/spi/project/SingleMethod.java:70]: I'd take the `file` from `NestedClass`.

File: ide/projectapi/src/org/netbeans/spi/project/SingleMethod.java [hunk 20]
```diff
@@ -31,6 +31,7 @@ public final class SingleMethod {
 
     private FileObject file;
     private String methodName;
+    private NestedClass nestedClass;
```
Comment (sdedic) [inline ide/projectapi/src/org/netbeans/spi/project/SingleMethod.java:34]: maybe final, as well as the other fields.

File: ide/projectapi/src/org/netbeans/spi/project/SingleMethod.java [hunk 21]
```diff
@@ -94,14 +126,15 @@ public boolean equals(Object obj) {
...
-        return other.file.equals(file) && other.methodName.equals(methodName);
+        return other.file.equals(file) && other.methodName.equals(methodName) && other.nestedClass.equals(nestedClass);
     }
 
     @Override
     public int hashCode() {
         int hash = 7;
         hash = 29 * hash + this.file.hashCode();
         hash = 29 * hash + this.methodName.hashCode();
+        hash = 29 * hash + this.nestedClass.hashCode();
```
Comment (sdedic) [inline ide/projectapi/src/org/netbeans/spi/project/SingleMethod.java:137]: :warning: Objects.hashCode() to avoid NPE on nestedClass; similar for equals. Unlike the other fields, nestedclass is optional.

File: java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/debugging/launch/NbLaunchDelegate.java [hunk 22]
```diff
@@ -632,11 +656,27 @@ static Lookup createTargetLookup(Project prj, SingleMethod singleMethod, FileObj
...
+        }
         if (toRun != null) {
             arr.add(toRun.getLookup());
         }
         return new ProxyLookup(arr.toArray(new Lookup[0]));
     }
+    
+    static ContainedProjectFilter getProjectFilter(Project prj, Map<String, Object> launchArguments) {
+        List<String> projectsArg = argsToStringList(launchArguments.get("projects"));
+        List<Project> projects = ProjectUtils.getContainedProjects(prj, false).stream()
```
Comment (sdedic) [inline java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/debugging/launch/NbLaunchDelegate.java:675]: Note: this does not allow to specify projects nested more levels in the project's hierarchy. Not a bug, and can be enhanced later in a compatible way.

File: java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/protocol/WorkspaceServiceImpl.java [hunk 23]
```diff
@@ -401,11 +403,15 @@ public CompletableFuture<Object> executeCommand(ExecuteCommandParams params) {
                     LOG.log(Level.INFO, "Project {2}: {0} test roots opened in {1}ms", new Object[] { testRoots.size(), (System.currentTimeMillis() - t), file});
                     BiFunction<FileObject, Collection<TestMethodController.TestMethod>, Collection<TestSuiteInfo>> f = (fo, methods) -> {
                         String url = Utils.toUri(fo);
+                        Project owner = FileOwnerQuery.getOwner(fo);
+                        String moduleName = owner != null ? ProjectUtils.getInformation(owner).getDisplayName(): null;
```
Comment (sdedic) [inline java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/protocol/WorkspaceServiceImpl.java:407]: Careful here. Check with customized gradle and maven project (display) names (different from artfiact names) that it works well.

File: java/java.lsp.server/vscode/src/testAdapter.ts [hunk 24]
```diff
@@ -63,53 +87,122 @@ export class NbTestAdapter {
...
+                        } else {
+                            await commands.executeCommand(request.profile?.kind === TestRunProfileKind.Debug ? COMMAND_PREFIX + '.debug.test': COMMAND_PREFIX + '.run.test', workspaceFolder.uri.toString());
+                        }
                     }
                 }
             }
             if (this.started) {
                 this.itemsToRun.forEach(item => this.set(item, 'skipped'));
+            } 
+            // TBD - message
```
Comment (sdedic) [inline java/java.lsp.server/vscode/src/testAdapter.ts:123]: TBD - for the future ?

File: java/java.lsp.server/vscode/src/testAdapter.ts [hunk 25]
```diff
@@ -63,53 +87,122 @@ export class NbTestAdapter {
...
+                item.children.forEach(child => this.set(child, state, message, noPassDown, dispatchBuildFailEvent));
             }
         }
     }
+    
+    dispatchTestEvent(state: SuiteState, testItem: TestItem): void {
+        if (testItem.parent && testItem.children.size > 0) {
+            this.dispatchEvent({
+                name: testItem.id,
+                moduleName: testItem.parent?.id,
```
Comment (sdedic) [inline java/java.lsp.server/vscode/src/testAdapter.ts:168]: nitpick: unnecessary parent?. as parent was checked to be non-falsy above.

File: java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/protocol/WorkspaceServiceImpl.java (hunk 23 above)
Comment (petrovic-d) [reply java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/protocol/WorkspaceServiceImpl.java:407]: Yes. For the maven we will have `parentName-artifactId`, and for the gradle it's `parentName:moduleName`

File: java/java.lsp.server/vscode/src/testAdapter.ts (hunk 24 above)
Comment (petrovic-d) [reply java/java.lsp.server/vscode/src/testAdapter.ts:123]: Yes. This case will happen when the build fails - tests are not even started. We need to decide what will be the message, or to somehow parse output from the console

File: ide/projectapi/src/org/netbeans/spi/project/NestedClass.java [hunk 26]
```diff
@@ -0,0 +1,147 @@
...
+ * If we have following structure: ParentClass (parent-of) ChildClass1 (parent-of) ChildClass2,
+ * for ChildClass1 className field would contain "ChildClass1" and topLevelClassName would contain "ParentClass",
+ * for ChildClass2 className field would contain "ChildClass1.ChildClass2" and topLevelClassName would contain "ParentClass"
+ * </p>
+ * 
+ * @author Dusan Petrovic
+ * 
+ * @since 1.99
+ */
+public final class NestedClass {
```
Comment (lahodaj) [inline ide/projectapi/src/org/netbeans/spi/project/NestedClass.java:39]: Not really a request to change anything, just an observation: this whole angle around `NestedClass` seems mostly orthogonal to the run tests in parallel feature, and could probably have been a separate PR.

**Review by sdedic — APPROVED** [review]
Looks good, thanks!

**Comment by sdedic** [comment]
@petrovic-d assuming rebase & squash

**Comment by matyasberry** [comment]
Hi, any chance of reusing these changes to also support what https://github.com/apache/netbeans/pull/3995 was intended to do?

**Comment by petrovic-d** [comment]
> Hi, any chance of reusing these changes to also support what #3995 was intended to do?

@matyasberry Hi, this PR should also add support for nested classes -> see [this](https://github.com/petrovic-d/netbeans/blob/40bae1af2279a038afbc2ce873ac0ffb7d5249d4/ide/projectapi/src/org/netbeans/spi/project/SingleMethod.java#L35)
